pkg/types/azure/machinepool.go

```diff
@@ -30,6 +30,13 @@ type MachinePool struct {
 	//
 	// +optional
 	UltraSSDCapability string `json:"ultraSSDCapability,omitempty"`
+
+	// AcceleratedNetworking specifies whether to enable accelerated networking
+	//  Accelerated networking enables single root I/O virtualization (SR-IOV) to a VM, greatly improving its
+	//  networking performance.
+	//
+	// +optional
+	AcceleratedNetworking bool `json:"acceleratedNetworking,omitempty"`
```

Comment:
I suggest that we do not use boolean for API types. It looks us into only being able to support two values. If Azure comes along with a new networking type, say Super Accelerated Networking, then we would need to add another field dependent on this field. Also, using a boolean makes it difficult to use a different default value.

I suggest a field that supports values "Basic" and "Accelerated". Do we want the default to be "Accelerated"? Do we support instance types that do not support "Accelerated"? If not, then we are missing some validation that the instance type selected supports accelerated networking.

Reply:
> I suggest that we do not use boolean for API types. It looks us into only being able to support two values. If Azure comes along with a new networking type, say Super Accelerated Networking, then we would need to add another field dependent on this field. Also, using a boolean makes it difficult to use a different default value.

I'm not tied to this API type. If you prefer a string we can do that as long as we interpret that correctly and pass the `true` value to the terraform. When I originally implemented this I attempted to match the API type from MS which is a `*bool` (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_types.go#L104). The feedback I received was this was not acceptable to place in our API. For this reason I went with a plain boolean as that made sense to me.

> I suggest a field that supports values "Basic" and "Accelerated". Do we want the default to be "Accelerated"? Do we support instance types that do not support "Accelerated"? If not, then we are missing some validation that the instance type selected supports accelerated networking.

Yes, we do want this to default to `true` or `accelerated` on the control plane. This is a performance improvement that we get for free on machines that have >= 4 vCPUs. Due to the nature of golang and defaulting variables to false this becomes more difficult to default to `true`. Maybe that changes if we change the type to a string?

This feature has a bit of history behind it. The upstream Azure cluster-api-provider has had support for this feature for years now. The provider has a bunch of features that we have not adopted and we are now moving from MAPI to CAPI. In the upstream provider a call occurs that fetches whether a machine has `AcceleratedNetworking` as a capability ( https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/9075098c210f1cd7e13db851813a15e232e2b13d/azure/services/resourceskus/sku.go#L73-L88). We do not have this functionality currently but will as soon as we adopt the upstream CAPI for Azure. 

The validation we had in place was to check the vmsize against a predetermined list of supported VM types. I know things are changing so I'll defer to the machine api team if they have any advice. I do know the API will fail with a detailed message that tells the users the instance type is unsupported. 
@elmiko @JoelSpeed

Reply:
i'm just doing a little background research on this, but i think @staebler has a good point from an api cleanliness perspective. will leave another comment once i have some details.

Reply:
i did a little background reading and the [azure provider spec](https://github.com/openshift/api/blob/master/machine/v1beta1/types_azureprovider.go#L115) already has this as boolean, but that doesn't mean the machinepool object needs to replicate that interface.

i think it's fine to change this to a string as long as the provider spec ends up being a boolean.

Reply:
> The validation we had in place was to check the vmsize against a predetermined list of supported VM types. I know things are changing so I'll defer to the machine api team if they have any advice. I do know the API will fail with a detailed message that tells the users the instance type is unsupported.

The installer can fetch the capabilities of the instance type to determine whether it supports accelerated networking. Similar validation was added for Ultra SSD support recently in https://github.com/openshift/installer/pull/5759.